tensorflow/python/keras/layers/preprocessing/preprocessing_stage.py

```diff
@@ -94,3 +96,170 @@ def map_fn(x):
+     'x2': TensorSpec(shape=(1,), dtype=tf.float32, name=None)}
+    >>> stage.adapt(ds)
+    >>> data_np = {'x1': np.ones((4, 5)), 'x2': np.ones((4, 1))}
+    >>> stage.adapt(data_np)
+
+    """
+    if not isinstance(data, dataset_ops.Dataset):
+      data = self._flatten_to_reference_inputs(data)
+      if any([not isinstance(datum, (np.ndarray, ops.EagerTensor))
+              for datum in data]):
+        raise ValueError(
+            '`adapt()` requires a batched Dataset, a list of EagerTensors '
+            'or Numpy arrays as input, got {}'.format(type(data)))
+      ds_input = [dataset_ops.Dataset.from_tensor_slices(x).batch(1)
+                  for x in data]
+
+    if isinstance(data, dataset_ops.Dataset):
+      # Validate the datasets to try and ensure we haven't been passed one with
+      # infinite size. That would cause an infinite loop here.
+      if tf_utils.dataset_is_infinite(data):
+        raise ValueError(
+            'The dataset passed to `adapt()` has an infinite number of '
+            'elements. Please use dataset.take(...) to make the number '
+            'of elements finite.')
+      # Unzip dataset object to a list of single input dataset.
+      ds_input = _unzip_dataset(data)
+
+    # Dictionary mapping reference tensors to datasets
+    ds_dict = {}
+    tensor_usage_count = self._tensor_usage_count
+    for x, y in zip(self.inputs, ds_input):
+      x_id = str(id(x))
+      ds_dict[x_id] = [y] * tensor_usage_count[x_id]
+
+    nodes_by_depth = self._nodes_by_depth
+    depth_keys = sorted(nodes_by_depth.keys(), reverse=True)
+
+    def build_map_fn(node, args, kwargs):
+      if not isinstance(args.element_spec, tuple):
+        def map_fn(*x):
+          return nest.flatten(node.layer(*x, **kwargs))
+      else:
+        def map_fn(*x):
+          return nest.flatten(node.layer(x, **kwargs))
+      return map_fn
+
+    for depth in depth_keys:
+      for node in nodes_by_depth[depth]:
+        # Input node
+        if node.is_input:
+          continue
+
+        # Node with input not computed yet
+        if any(t_id not in ds_dict for t_id in node.flat_input_ids):
+          continue
+
+        args, kwargs = node.map_arguments(ds_dict)
+        args = dataset_ops.Dataset.zip(nest.list_to_tuple(*args))
+
+        if hasattr(node.layer, 'adapt'):
+          node.layer.adapt(args, reset_state=reset_state)
+
+        map_fn = build_map_fn(node, args, kwargs)
+        outputs = args.map(map_fn)
+        outputs = _unzip_dataset(outputs)
+
+        # Update ds_dict.
+        for x_id, y in zip(node.flat_output_ids, outputs):
+          ds_dict[x_id] = [y] * tensor_usage_count[x_id]
+
+
+def _unzip_dataset(ds):
+  """Unzip dataset into a list of single element datasets.
+
+  Arguments:
+    data: A Dataset object.
+
+  Return:
+    A list of Dataset object, each correspond to one of the
+    `element_spec` of the input Dataset object.
+
+  Example:
+  >>> ds1 = tf.data.Dataset.from_tensor_slices([1, 2, 3])
```

Comment:
Every doctest requires a newline before and after it. Please follow this: https://www.tensorflow.org/community/contribute/docs_ref

Here and everywhere else.

Reply:
@yashk2810 Thank you for the comments. I have made those changes.